backfiller/backfiller/main.py

```diff
@@ -129,159 +114,308 @@ def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+		"""Backfill from remote node.
+	
+		Backfill from node/stream/variants to base_dir/stream/variants for each
+		hour in hours.
+	
+		Keyword arguments:
+		recent_cutoff -- Skip backfilling segments younger than this number of
+			seconds to prioritise letting the downloader grab these segments."""
+	
+		for variant in self.variants:
+	
+			for hour in hours:
+	
+				self.logger.info('Backfilling {}/{}'.format(variant, hour))
+	
+				local_segments = set(list_local_segments(self.base_dir, self.stream, variant, hour))
+				remote_segments = set(list_remote_segments(self.node, self.stream, variant, hour))
+				missing_segments = list(remote_segments - local_segments)
+	
+				# randomise the order of the segments to reduce the chance that
+				# multiple workers request the same segment at the same time
+				random.shuffle(missing_segments)
+	
+				for missing_segment in missing_segments:
 	
-	backfill(base_dir, stream, variants, 3, nodes=nodes)
+					if self.stopping.is_set():
+						return
 
-	backfill(base_dir, stream, variants, nodes=nodes)
+					path = os.path.join(self.stream, variant, hour, missing_segment)
 	
-	# I'm sure there is a module that does this in a more robust way 
-	# but I understand this and it gives the behaviour I want
-	while True:
+					# test to see if file is a segment and get the segments start time
+					try:
+						segment = common.parse_segment_path(path)
+					except ValueError as e:
+						self.logger.warning('File {} invaid: {}'.format(path, e))
+						continue
+	
+					#to avoid getting in the downloader's way ignore segments less than recent_cutoff old
+					if datetime.datetime.utcnow() - segment.start < datetime.timedelta(seconds=recent_cutoff):
+						self.logger.debug('Skipping {} as too recent'.format(path))
+						continue
+	
+					get_remote_segment(self.base_dir, self.node, self.stream, variant, hour, missing_segment)
+				self.logger.info('{} segments in {}/{} backfilled'.format(len(missing_segments), variant, hour))
+
+	def run(self):
+		self.logger.info('Starting')
+
+		while not self.stopping.is_set():
+
+			try:
+				self.backfill(list_hours(self.node, self.stream, self.variants, self.start))
+				self.failures = 0 #reset failure count on a successful backfill
+				if not self.run_once:
+					self.stopping.wait(common.jitter(self.WAIT_INTERVAL))
+
+			except Exception:
+				self.failures += 1
+				delay = common.jitter(self.RETRY_INTERVAL * 2**self.failures)
+				self.logger.exception('Backfill failed. Retrying in {:.0f} s'.format(delay))
+				self.stopping.wait(delay)
+
+			if self.run_once:
+				break
+		
+		self.logger.info('Worker stopped')
+		self.done.set()
+		if self.node in self.manager.workers:
+			del self.manager.workers[self.node]
 
-		now = datetime.datetime.now()
+							
+def main(base_dir='.', streams='', variants='', metrics_port=8002,
+	static_nodes='', backdoor_port=0, start=None, run_once=False,
+	node_file=None, node_database=None, localhost=None):
+	"""Backfiller service.
```

Comment:
`variants` should default to `source` like downloader.

assuming we go with having localhost refer to the node's "name", it would make sense to default to `socket.gethostname()`.

`streams` should probably be a positional arg, not an option `--streams`, since it's almost always required (backfilling no streams is possible but not useful).
Ideally, it should work the same way as downloader and can be specified as multiple positional args, ie. `backfiller --blah=blah channel1 channel2 channel3`. You can do this by making `streams` the first arg, not giving it a default, and adding:
```
@argh.arg("streams", nargs="*")
```
the `nargs="*"` tells it that the positional arg may be given 0 or more times (also possible: `?` for optional, `N` for exactly n times (eg. `1`, which is the default), or `+` for "1 or more").

Reply:
Done